File: source/Stmt.java
import java.util.*;

abstract class Stmt {
    interface Visitor <R> {
        R visitBlockStmt(Block stmt);
        R visitClassStmt(Class stmt);
        R visitExpressionStmt(Expression stmt);
    }

    static class Block extends Stmt {
        Block(List<Stmt> statements) {
            this.statements = statements;
        }

        <R> R accept(Visitor<R> visitor) {
            return visitor.visitBlockStmt(this);
        }

        final List<Stmt> statements;
    }

    static class Class extends Stmt {
        Class(Token name) {
            this.name = name;
        }

        <R> R accept(Visitor<R> visitor) {
            return visitor.visitClassStmt(this);
        }

        final Token name;
    }

    static class Expression extends Stmt {
        Expression(Expr expression) {
            this.expression = expression; 
        }

        <R> R accept(Visitor<R> visitor) {
            return visitor.visitExpressionStmt(this);
        }

        final Expr expression;
    }
}

File: source/Stmt.kt
sealed class Stmt {
    class Class(
        val name: Token,
        val constructor: Constructor,
        val methods: List<Function>,
        val fields: Map<Token, Expr>) : Stmt()

    class If(val condition: Expr, val thenBranch: Stmt, val elseBranch: Stmt?): Stmt()

    class Print(val expression: Expr): Stmt()

    class Expression(val expression: Expr) : Stmt()

    class Constructor(
        val fields: Map<Int, Token>,
        val parameters: List<Token>,
        val constructorBody: Expr.Block
    ) : Stmt()
    class Function(val name: Token, val functionBody: Expr.Function): Stmt()

    class Var(val name: Token, val initializer: Expr): Stmt()

    class While(val condition: Expr, val body: Stmt): Stmt()
}